spark/src/main/java/org/apache/iceberg/spark/source/Writer.java

```diff
@@ -303,66 +305,128 @@ public String toString() {
+      private final int partitionId;
+      private final long taskId;
+      private final long epochId;
+
+      EncryptedOutputFileFactory(int partitionId, long taskId, long epochId) {
+        this.partitionId = partitionId;
+        this.taskId = taskId;
+        this.epochId = epochId;
+      }
+
+      private String generateFilename() {
+        return format.addExtension(String.format("%05d-%d-%s", partitionId, taskId, UUID.randomUUID().toString()));
+      }
+
+      /**
+       * Generates EncryptedOutputFile for UnpartitionedWriter.
+       */
+      public EncryptedOutputFile newOutputFile() {
+        OutputFile file = fileIo.newOutputFile(locations.newDataLocation(generateFilename()));
+        return encryptionManager.encrypt(file);
+      }
+
+      /**
+       * Generates EncryptedOutputFile for PartitionedWriter.
+       */
+      public EncryptedOutputFile newOutputFile(PartitionKey key) {
+        OutputFile rawOutputFile = fileIo.newOutputFile(locations.newDataLocation(spec, key, generateFilename()));
+        return encryptionManager.encrypt(rawOutputFile);
+      }
+    }
   }
 
   private interface AppenderFactory<T> {
     FileAppender<T> newAppender(OutputFile file, FileFormat format);
   }
 
-  private static class UnpartitionedWriter implements DataWriter<InternalRow>, Closeable {
+  private interface OutputFileFactory<T> {
+    T newOutputFile();
+    T newOutputFile(PartitionKey key);
+  }
+
+  private static class UnpartitionedWriter implements DataWriter<InternalRow> {
     private final FileIO fileIo;
-    private FileAppender<InternalRow> appender = null;
-    private Metrics metrics = null;
-    private List<Long> offsetRanges = null;
-    private final EncryptedOutputFile file;
+    private FileAppender<InternalRow> currentAppender = null;
+    private final OutputFileFactory<EncryptedOutputFile> fileFactory;
+    private final FileFormat format;
+    private final AppenderFactory<InternalRow> appenderFactory;
+    private EncryptedOutputFile currentFile = null;
+    private final List<DataFile> completedFiles = Lists.newArrayList();
+    private final long targetFileSize;
 
     UnpartitionedWriter(
-        EncryptedOutputFile outputFile,
+        OutputFileFactory<EncryptedOutputFile> fileFactory,
         FileFormat format,
-        AppenderFactory<InternalRow> factory,
-        FileIO fileIo) {
+        AppenderFactory<InternalRow> appenderFactory,
+        FileIO fileIo,
+        long targetFileSize) {
+      this.fileFactory = fileFactory;
+      this.format = format;
+      this.appenderFactory = appenderFactory;
       this.fileIo = fileIo;
-      this.file = outputFile;
-      this.appender = factory.newAppender(file.encryptingOutputFile(), format);
+      this.targetFileSize = targetFileSize;
+
+      openCurrent();
     }
 
     @Override
-    public void write(InternalRow record) {
-      appender.add(record);
+    public void write(InternalRow record) throws IOException {
```

Comment:
This is called in a tight loop and `currentAppender.length()` is expensive in some cases. Parquet will find the size of the current row group and add that to its position in the file. Can we update this to check every 1,000 records or so?

Reply:
Ah, I missed that in between all the abstraction. Makes sense, will fix.

Reply:
Fixed.

Reply:
Upon further testing, we have found that we are missing the target-file-size by a whole row-group size.

Taking a look at `ParquetWriter`, the `currentAppender.length()` maps to:

https://github.com/apache/incubator-iceberg/blob/776210e3e49591a47c5d3d93ecd6d5c8b22e19b5/parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java#L126-L133

And this calls `ParquetFileWriter` from parquet-hadoop:
```java
  /**
   * @return the current position in the underlying file
   * @throws IOException if there is an error while getting the current stream's position
   */
  public long getPos() throws IOException {
    return out.getPos();
  }
```

`out` being an OutputStream. So it doesn't look like we are doing the estimation of length including the current row group.

Seems like `ParquetWriter` should really be doing `writer.getPos() + writeStore.getBufferedSize()` ?

Reply:
I agree

Reply:
Fixed.